# LACONv
A large copy number variation detection algorithm using read-depth of coverage in targeted panels of genes

Currently next-generation sequencing technologies are used to scan the genome for SNPs, indels, and structural variants that range from targeted genes in custom panels to whole exome/genome wide studies. Even though structural variations, such as copy number variations, which includes insertions, deletions and duplications are routinely identified in clinical domain by array CGH or MLPA, there are no “best practice guidelines” in the NGS community to assess the validity of CNV detection in data sequenced from NGS neither the clinical study of the methods. Several computational tools for CNV detection have been developed in the last years. Frequently, they are exome-oriented methods, or they are specialized in a particular NGS platform. This issue motivated us to develope an approach adequate to identify CNVs in different kinds of panel of genes and sequencing platforms. We here present a methodology for CNV detection indicated for tailored panel of candidate genes including whole-exome data. This methodology has been validated on large amount of Breast cancer, Mody diabetes, Cardiopathy syndromes, Dravet Syndrome, or Skeletal Dysplasia samples.
